### PR TITLE
added a message that data will be lost if the user leaves or reload t…

### PR DESCRIPTION
**Pull request by pospyra**
https://github.com/BinaryStudioAcademy/bsa-2023-leetwars/assets/68563967/a0914f99-e67f-445a-93e3-68e77d35b0bc

image
(default text there)

https://trello.com/c/4sOpMteM/168-bug58-create-challenge-page-all-data-is-lost-when-the-page-is-refreshed

### CODE REVIEW

**Review by AyTee001 — APPROVED** [review]

File: frontend/src/app/core/guards/unsaved-changes.guard.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    providedIn: 'root',
+})
+export class UnsavedChangesGuard<T extends HasUnsavedChanges> implements CanDeactivate<T> {
+    constructor(
+        private modalService: NgbModal,
+    ) {}
+
+    canDeactivate(component: T): Promise<boolean> {
+        return new Promise<boolean>((resolve) => {
+            if (component.unsavedChanges) {
```
Comment (tatianahutii) [inline frontend/src/app/core/guards/unsaved-changes.guard.ts]:
```suggestion
            if (!component.unsavedChanges) {
                resolve(true);
            }
```

File: frontend/src/app/core/guards/unsaved-changes.guard.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                        },
+                    },
+                    {
+                        text: 'Cancel',
+                        handler: () => {
+                            resolve(false);
+                            modalRef.close();
+                        },
+                    },
+                ];
```
Comment (tatianahutii) [inline frontend/src/app/core/guards/unsaved-changes.guard.ts]: to a separate method

File: frontend/src/app/core/guards/unsaved-changes.guard.ts (hunk 1 above)
Comment (tatianahutii) [reply frontend/src/app/core/guards/unsaved-changes.guard.ts]: or even better to use ternary operator here

**Review by tatianahutii — APPROVED** [review]

**Review by tatianahutii — CHANGES_REQUESTED** [review]
See the previous comments

File: frontend/src/app/core/guards/unsaved-changes.guard.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    constructor(
+        private modalService: NgbModal,
+    ) {}
+
+    canDeactivate(component: T): Promise<boolean> {
+        return new Promise<boolean>((resolve) => {
+            resolve(component.unsavedChanges
+                ? this.showConfirmationModal()
+                : true);
+        });
```
Comment (tatianahutii) [inline frontend/src/app/core/guards/unsaved-changes.guard.ts]:
```suggestion
        return component.unsavedChanges
            ? this.showConfirmationModal()
           :  Promise.resolve(true);
        });
```

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by tatianahutii — APPROVED** [review]

**Pull request merged**